Reorganization/adjustments to helpers machinery

eteq opened the pull request:
This is in part a follow-on to #149 and in others a continuation of the ideas from #104 - it's sort of a grab-bag of things I noticed while playing with the new specviz helper object:

1. It re-works the base helper class to better reflect ideas from #104 and #133 : specifically that it's really the *application* object that the helper is tied to, rather than a specific configuration.  
2. makes `app` in the base helper public - note that this is critical to the scheme from #104 - the user has to be free to interact with the `Application` object when they want to do "glue-y" things, while still carrying around just a reference to the helper object.
3. Moves the base helper into `jdaviz.core`
4. Creates helper objects in each of the specviz, mosviz, and cubeviz config directories
5. changes the base-level imports of `jdaviz` to include the helpers and `Application` on the theory that this is one the users most want if they don't really want to understand the details.
6. Renames the `app` in `TemplateMixin` to `glueapp` to reduce confusion.

Note that 6 is the one I'm most uncertain about because I don't really understand what it's meant to be used for (@nmearl may have some insight?)

cc @dtn5ah @nmearl

rosteen commented:
Not going to dig in at 9PM, but my initial guess on (6) is that it's the change @nmearl made to enable the plugins to access the overarching app instance (so they can e.g. get to viewer data via the app methods). Could be totally off base.

eteq commented:
That was the most satistrating (satisfying while also frustrating) commit I've done in a while: a commit adding a single new line for PEP8 compliance after all that moving stuff around...

nmearl commented on jdaviz/core/helpers.py:
```diff
@@ -0,0 +1,31 @@
+    _default_configuration = 'default'
+
+    def __init__(self, app=None):
+        from ..app import Application  # inside to prevent possible circular imports
```
Are you anticipating circular imports, or did you actually run into one?

eteq replied:
Anticipating.  Which now that you bring it up... is probably something I should avoid until/if it actually happens, so I'll move it to top-level.

nmearl commented on jdaviz/core/template_mixin.py:
```diff
@@ -1,26 +1,34 @@
-        app = kwargs.pop('app', None)
         obj = super().__new__(cls, *args, **kwargs)
-        setattr(obj, '_app', app)
+        obj._app = kwargs.pop('app', None)
```
We want to pop `app` before the `kwargs` dict is passed to `super` -- `app` can cause issues when passed to parent classes.

eteq replied:
Ah, gotcha. I meant this really to just change the setattr but that totally makes sense, my bad, will fix.

nmearl commented on jdaviz/core/template_mixin.py:
```diff
@@ -1,26 +1,34 @@
+        return self._app
+
+    @property
+    def glueapp(self):
```
This is never true. When [plugins are instantiated](https://github.com/spacetelescope/jdaviz/blob/master/jdaviz/app.py#L853), they are passed `self` as the `app` reference. Therefore, the `TemplateMixin` would be passing `app=SpecViz()`, or `app=CubeViz()` depending on context, of `app=Application()` with no configuration. In any case, it's never passed the glue application instance. The only accessor to the underlying glue app should exist on the JDAViz `Application` class in `app.py`.

eteq replied:
Aha, that's my misunderstanding.  My main reason for making this change is that I think having two layers of `app` is confusing, so I'm thinking this should have a more specific name than `app` (or alternatively, the `helper` could have it's name changed...).  What do you think is a more explicit term for this?

(Also, can you use `suggest` in Github here to add a short docstring saying what this was intended here?)

nmearl replied:
Currently, there is no accessor for the underlying glue application, so there's nominally only one "app" and that's the current JDAViz application the user is interacting with. I think instead we should add a property to the `app.py` file called `glue_app` or something.

eteq replied:
I'm still confused here @nmearl: my intent behind this was to not have an attribute with an ambiguous name `app` refer to two different *classes*/APIs. Since`SpecViz` and `Application` definitely have different APIs, I'm trying to separate them out into two different attribute names.  I see your point that the link to the *glue* application should live in the `Application`, but the fact that I was confused about this is illustrating the problem I'm trying to solve... so are you saying I should revert this? If so, what should we say as the docstring of which meaning of "app" the `TemplateMixin` is?

nmearl commented on jdaviz/core/template_mixin.py:
```diff
@@ -1,26 +1,34 @@
 
     @property
     def app(self):
+        warn(DeprecationWarning('The usage `.app` from a jdaviz application is'
+                                ' deprecated and should be replaced by '
+                                '`.glueapp` to prevent confusion with other '
+                                'meanings of "app"'))
+        return self._app
+
+    @property
+    def glueapp(self):
```
Having thought about this a bit more, I misspoke. This would not be access to the `SpecViz` or `CubeViz` helper class instance, but instead the parent jdaviz `Application`, since the plugin instantiate happens there and the helper classes are not subclasses.

```suggestion
        """
        Allows access to the underlying jdaviz application instance. This is **not** access to the helper 
        class, but instead the core application.
        """
```

eteq replied:
Oops!  This message crossed with the one I just posted - yes, this makes sense now, thanks for clearing that up!

eteq commented:
Alright I think I've addressed all the comments here.  I also added one more change based on an out-of-band conversation with @nmearl: I changed the `Application` to inherit from `VuetifyTemplate` and `glue.core.HubListener` because the `TemplateMixin` leads to some confusion with the `app` attribute, as `Application` shouldn't have an `app` since it *is* the `app`.
